service/topology_coordinator.cc

```diff
@@ -2597,17 +2597,27 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
 };
 
 future<std::optional<group0_guard>> topology_coordinator::maybe_migrate_system_tables(group0_guard guard) {
+    const auto view_builder_version = co_await _sys_ks.get_view_builder_version();
+    if (view_builder_version == db::system_keyspace::view_builder_version_t::v2) {
```

Comment:
This `if` is one part of the solution, but it is not enough - you need to make sure that the topology coordinator is notified when all nodes become alive. It should be enough to notify the topology state machine condition variable.

I was looking into this issue previously actually, and this might not be that trivial - although the topology coordinator is already an endpoint lifecycle subscriber and you could potentially use the `on_up` notification, I am not sure whether all appropriate data structures will be updated and the query layer will know that all nodes are UP at the point when you perform the CL=ALL query.

Reply:
Besides, why not just put the `else if` part inside the `if (view_builder_version == db::system_keyspace::view_builder_version_t::v1_5) {` block?

Reply:
> I've tried to follow the update path but I'm not very familiar with this code. But isn't the notification in on_up() + !get_dead_nodes().empty() check enough?

I'm not sure. `get_dead_nodes()` looks at the topology state machine and the gossiper state (calls `gossiper.is_alive`). If it is guaranteed that the querying code uses the same check, or uses some other data structure that is guaranteed to be updated before the `on_up` callback in the topology coordinator runs, then it's fine - otherwise you risk that the check is flaky and migration will sometimes not run, maybe until some unrelated event triggers the condition variable.

Reply:
@kbr-scylla maybe you know the answer?